### PR TITLE
Clean up numeric.h and split FaceFositionCache from it

### PR DESCRIPTION
**Pull request by ShadowNinja**
This allows me to remove a lot of includes from numeric.h, which is included by ~127 sources, reducing compile time.
Also:
- I optimized FacePositionCache by removing a map lookup and vector copy from both branches of getFacePosition.
- I fixed `myfloor(0.0)` returning `-1`.

### CODE REVIEW

File: src/mg_ore.cpp [hunk 1]
```diff
@@ -20,9 +20,10 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "mg_ore.h"
 #include "mapgen.h"
 #include "noise.h"
-#include "util/numeric.h"
 #include "map.h"
 #include "log.h"
+#include <algorithm>
```
Comment (est31) [inline src/mg_ore.cpp]: why is this needed?

File: src/util/numeric.h [hunk 2]
```diff
@@ -347,16 +305,18 @@ class IntervalLimiter
...
 			return false;
-		m_accumulator -= wanted_interval;
+		accumulator -= wanted_interval;
 		return true;
 	}
-protected:
-	float m_accumulator;
+
+private:
+	float accumulator;
```
Comment (est31) [inline src/util/numeric.h]: It should still be named m_accumulator. Its a private member.

File: src/face_position_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
+/*
+Minetest
+Copyright (C) 2010-2015 celeron55, Perttu Ahola <celeron55@gmail.com>
```
Comment (est31) [inline src/face_position_cache.cpp]: I think this particular class was implemented by @nerzhul .

File: src/util/numeric.h [hunk 4]
```diff
@@ -24,42 +24,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 #include "../irr_v3d.h"
 #include "../irr_aabb3d.h"
-#include "../threading/mutex.h"
-#include <list>
-#include <map>
-#include <vector>
-#include <algorithm>
 
+#define ARRLEN(a) (sizeof(a) / sizeof((a)[0]))
+#define MYMIN(a,b) ((a)<(b)?(a):(b))
```
Comment (est31) [inline src/util/numeric.h]: space

File: src/util/numeric.h [hunk 5]
```diff
@@ -24,42 +24,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+#define MYMAX(a,b) ((a)>(b)?(a):(b))
```
Comment (est31) [inline src/util/numeric.h]: here too

File: src/util/numeric.h [hunk 6]
```diff
@@ -24,42 +24,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+#define rangelim(d, min, max) ((d) < (min) ? (min) : ((d)>(max)?(max):(d)))
+#define myfloor(x) ((x) > 0.0 ? (int)(x) : (int)(x) - 1)
+// Requires <algorithm>
+#define CONTAINS(c, v) (std::find((c).begin(), (c).end(), (v)) != (c).end())
+// The naive swap performs better than the xor version
+#define SWAP(t, x, y) do { \
```
Comment (est31) [inline src/util/numeric.h]: We have agreed to manually align these backslashes with spaces.

File: src/mg_ore.cpp (hunk 1 above)
Comment (ShadowNinja) [inline src/mg_ore.cpp]: `CONTAINS` requires it.  It's not included by `numerc.h` because only a few files actually use it and we don't want to require the compiler to load it for avarything that includes `numeric.h`.

File: src/util/numeric.h (hunk 5 above)
Comment (ShadowNinja) [inline src/util/numeric.h]: I just moved it.

File: src/util/numeric.h (hunk 4 above)
Comment (ShadowNinja) [inline src/util/numeric.h]: I just moved it.

File: src/util/numeric.h (hunk 6 above)
Comment (ShadowNinja) [inline src/util/numeric.h]: That's impossible to do properly with tab indentation, and I never heard of such an agreement.

File: src/mg_ore.cpp (hunk 1 above)
Comment (est31) [inline src/mg_ore.cpp]: Perhaps add a comment then?

File: src/util/numeric.h (hunk 4 above)
Comment (est31) [inline src/util/numeric.h]: Its noted as + on the diff, and blame will show your name.

File: src/util/numeric.h (hunk 5 above)
Comment (est31) [inline src/util/numeric.h]: Its noted as + on the diff, and blame will show your name.

File: src/util/numeric.h (hunk 6 above)
Comment (est31) [inline src/util/numeric.h]: Yes it is impossible, take what width tabs show for you: https://forum.minetest.net/viewtopic.php?f=3&t=12247

Comment (Zeno-) [inline src/util/numeric.h]: Hopefully x never == 0.0 for myfloor()

Comment (ShadowNinja) [inline src/util/numeric.h]: Fixed.

File: src/mg_ore.cpp (hunk 1 above)
Comment (ShadowNinja) [inline src/mg_ore.cpp]: https://github.com/minetest/minetest/pull/3256/files#diff-a408137ced7567bc503aa6f6386af956R34

**Comment by kwolekr** [comment]
I think before refactoring FacePositionCache a second look at whether or not it's actually needed should be made.
The best code is code that doesn't exist.

**Comment by est31** [comment]
@kwolekr +1.

**Comment by ShadowNinja** [comment]
@kwolekr: Acording to 7c8793cbea1ea83109b7d9d6974d3f6991efcec8 it makes find_node_near substantially faster.

**Comment by kwolekr** [comment]
Not that I don't trust nrz, it's just that I'd like the actual performance improvement to be independently verified...

**Comment by nerzhul** [comment]
@kwolekr no problem, take a callgrind, mesure yourself :). Generally i don't like caches, except when we are doing thousand times the same thing

**Comment by nerzhul** [comment]
@ShadowNinja can you rebase this PR ?

**Comment by ShadowNinja** [comment]
@nerzhul: Done.

**Comment by ShadowNinja** [comment]
0adc4823309fb29c57d349e301ff7bb11c034c4b 723ab1a14a4a61486d31db022eb62794626cdf2a
